### ISSUE TITLE
UniRig: hosted encoder.onnx outputs 100% NaN latents — --algo unirig has never run the ML model

### ISSUE THREAD

**Issue by fernandotonon** [fernandotonon/QtMeshEditor#1025]
The hosted UniRig encoder export produces **100% NaN latents** on every input. `--algo unirig` has therefore never actually run the ML model — every call silently falls back to the Pinocchio template rig.

Found while starting #1013 (non-human categories). It blocks that issue, and any other UniRig work.

## Evidence

Encoder inputs are clean; the encoder output is structurally correct and entirely non-finite:

```
[unirig] enc.pts   n=98304   nonfinite=0  range=[-0.9997, 0.9994]
[unirig] enc.nrm   n=98304   nonfinite=0  range=[-1.0000, 1.0000]
[unirig] encOut[0] 'latents' shape=1x1024x1024 n=1048576 nonfinite=1048576 first4=nan nan nan nan
[unirig] dead state: step=2 validTokens=260 haveAnyJoint=0 top5=2(nan),4(nan),0(nan),1(nan),3(nan)
```

The decode then dies at step 2 — not because the FSM is wrong (260 tokens are valid) but because every logit is NaN, so no token can be selected. That surfaces as the existing message *"UniRig: constrained decode reached a dead state."*

## What it is NOT

Each of these was tested and ruled out:

- **Not input preprocessing** — points normalise to [-1,1], normals are unit, zero non-finite values in either tensor.
- **Not CoreML** — `QTMESH_ONNX_FORCE_CPU=1` gives byte-identical all-NaN output, so it is not an EP numerical issue.
- **Not mesh-specific** — reproduced on a humanoid (`jana.obj`), a car (`buick_riviera*.glb`), i.e. it fails on the class the checkpoint was trained for, not just non-human input.
- **Not a truncated download** — local files match the hosted sizes exactly: encoder 216,674,173 / decoder 1,223,378,677 / embed 1,093,821 bytes.
- **Not a wiring/shape bug** — the output shape `1x1024x1024` is exactly the expected latent prefix.

So the fault is in the exported graph itself (`scripts/export-unirig-onnx.py`), not in `UniRigPredictor`.

## Impact

`AutoRig::Algorithm::UniRig` is selectable from the CLI (`--algo unirig`), MCP (`auto_rig {algo}`) and the Inspector's Algorithm picker. All three silently produce a **template** rig while reporting `algorithmUsed`/`fallbackReason`. The rig is still usable — the fallback works as designed — but nobody has ever received the ML result, which is the entire point of #408.

## Suggested next steps

1. Re-run `scripts/export-unirig-onnx.py` and **assert parity against torch** on a fixed input before uploading — the SkinTokens export (#819 Slice C) already does this (`~1e-5` per graph) and is the pattern to copy. The likely culprits are fp16/overflow in the attention stack or an unsupported op lowered to something numerically unstable.
2. Add a **cheap startup guard** in `UniRigPredictor`: if the encoder's first latents are non-finite, fail with "the UniRig encoder export is producing NaN — falling back" rather than the misleading downstream *"constrained decode reached a dead state"*, which points at the FSM.
3. Consider a tiny parity fixture in CI so a bad re-export cannot ship silently again.

I have not attempted a re-export here — it needs the upstream UniRig repo + torch, and should be validated against torch outputs rather than guessed at.

**Comment by fernandotonon** [comment]
## Follow-up: re-tested the "not full-body humanoids" hypothesis — it does not hold

Fair challenge, and it exposed a real weakness in my original evidence: **my "humanoid control" was not a full-body humanoid.** `jana.obj` is a head-and-shoulders bust. So both meshes I originally tested (bust + car) were outside UniRig's training distribution, and my claim that it "fails on the class the checkpoint was trained for" was not supported by what I had run.

Re-tested properly. The conclusion is unchanged, but now it is actually earned.

### 1. A genuine full-body humanoid

Took a Mixamo character (`Hip Hop Dancing.fbx`, 6,968 verts) and converted it to OBJ to strip the skeleton — an unrigged full-body humanoid, exactly UniRig's training case:

```
[unirig] latents 1024x1024 n=1048576 nonfinite=1048576 (100.0%)
```

### 2. A mathematically perfect synthetic input

Stronger still — replaced the sampled cloud with a Fibonacci unit sphere: all values finite, points inside [-0.9, 0.9], normals exactly unit length. Nothing about it can be blamed on mesh topology, scale, normals or sampling:

```
[unirig] SYNTHETIC unit-sphere input substituted
[unirig] latents 1024x1024 n=1048576 nonfinite=1048576 (100.0%)
```

### Summary

| input | non-finite latents |
|---|---|
| bust (`jana.obj`) | 1048576 / 1048576 |
| vehicle (`buick_riviera*.glb`) | 1048576 / 1048576 |
| **full-body humanoid** (Mixamo, unrigged) | **1048576 / 1048576** |
| **synthetic unit sphere** | **1048576 / 1048576** |

The encoder cannot produce a single finite value for **any** input, including one that is perfect by construction. Combined with the earlier eliminations (not CoreML — CPU is byte-identical; not a truncated download — sizes match the hosted files exactly; not a shape bug — output is the expected 1x1024x1024), the fault is in the exported graph itself.

So #1025 stands as written, and the re-export with torch-parity assertions is still the fix. Nothing about mesh type or "full body" changes it.

**Comment by fernandotonon** [comment]
## Post-mortem: the export was fine — the local download was corrupt

Hashed instead of theorised. Three same-size (216,674,173 B) encoder files, three different digests:

| file | sha256 | weights | latents |
|---|---|---|---|
| HF-hosted (LFS oid) | `857d2381…` | all finite | finite, input-sensitive |
| export stash (`~/unirig_onnx_export`) | `e419b3c6…` | all finite | finite (different baked FPS indices) |
| **the copy in `ai_models/unirig/` that produced this issue** | `f8d53eee…` | **7 + 261 NaN/±3e38 values in two `resblocks.2` matrices** | 100 % NaN |

The local copy differs from HF in 309,884 bytes within one ~8.6 MB region (offsets 39.4–48.0 MB). Every conclusion in the issue body holds *for that file*; the "the fault is in the exported graph" inference does not — the hosted graph passes the same probes cleanly. So the ML path never ran because a corrupted download was loaded and trusted, which is exactly the class of failure #1038's digest verification exists for — and no consumer passed a digest.

**Fix:** PR #1047 — `ModelFetch` verifies an already-present file against a known digest (cached per process) and deletes + re-fetches on mismatch; UniRig passes its three HF LFS oids. Live run on this machine: mismatch detected, encoder re-fetched, `algorithm: unirig`, 52 joints, exit 0 in 78 s — the first real UniRig skeleton here.

**Secondary finding, filed separately:** #1046 — the export froze farthest-point sampling at trace time (graph requires N ≥ 2048; larger clouds use a fixed index subset).
